twilight-cache-inmemory/src/traits.rs

```diff
@@ -0,0 +1,247 @@
+//! Traits for implementing a [`InMemoryCache`] with custom structs.
+//!
+//! By default, the cache uses widely compatible default types that contain almost all
+//! fields that are present in the Discord API. Fields that are never used by the user
+//! will result in excess memory usage that will especially matter to big bots with a
+//! lot of cached data.
+//!
+//! The traits in this module allow creating custom cached representations of Discord
+//! API models compatible with the [`InMemoryCache`]. They may be mixed with the default
+//! types provided by twilight, which also implement these traits.
+//!
+//! Many traits require getters for certain types, which means they are used for caching
+//! logic. However, users generally won't have to store all the fields. It is possible
+//! to return `None` or empty arrays on most of the methods if the data that is accessed
+//! is not stored in the custom implementation.
+//!
+//! [`InMemoryCache`]: crate::InMemoryCache
+
+use crate::model::member::ComputedInteractionMemberFields;
+use std::fmt::Debug;
+use twilight_model::{
+    application::interaction::InteractionMember,
+    channel::{
+        message::{Reaction, Sticker},
+        Channel, ChannelType, Message, StageInstance,
+    },
+    gateway::{
+        payload::incoming::{GuildUpdate, MemberUpdate, MessageUpdate},
+        presence::Presence,
+    },
+    guild::{Emoji, Guild, GuildIntegration, Member, PartialMember, Role},
+    id::{
+        marker::{ChannelMarker, GuildMarker, RoleMarker, StickerMarker, UserMarker},
+        Id,
+    },
+    user::{CurrentUser, User},
+    util::{ImageHash, Timestamp},
+    voice::VoiceState,
+};
+#[cfg(feature = "permission-calculator")]
+use twilight_model::{channel::permission_overwrite::PermissionOverwrite, guild::Permissions};
+
+/// Trait for a generic cached representation of a [`Member`].
+pub trait CacheableMember:
+    From<Member>
+    + From<(
+        Id<UserMarker>,
+        InteractionMember,
+        ComputedInteractionMemberFields,
```

Comment:
I think this type is private. And to catch this we should test having a custom impl of each trait.

Reply:
Any type implementing `CacheableMember` needs a `From<(.., ComputedInteractionMemberFields)>` but that's impossible as `ComputedInteractionMemberFields` is not public (impossible to name).

Reply:
Yes but I'm not sure if a test can catch that since tests can access private structs and fields if I'm not mistaken

Reply:
Oh I see what you mean. It could be done in integration tests (like #2228) or in an example, but honestly it's quite easily catchable so I'm okay with not holding it up for this (butthe trait still needs fixing!)

Reply:
I think I'll just extend the example to provide replacements for all types, that should catch it